src/value_generators/src/DocumentGenerator.cpp

```diff
@@ -1388,6 +1393,130 @@ class ArrayGenerator : public Generator<bsoncxx::array::value> {
     const UniqueGenerator<int64_t> _nTimesGen;
 };
 
+/** `{^DistinctArray: {of: {^RandomInt {start: 10, end: 1000}}}, number: 2}` */
+class DistinctArrayGenerator : public Generator<bsoncxx::array::value> {
+public:
+    // Make the variant type easier to use.
+    using GeneratorTypes = std::variant<
+        UniqueGenerator<bsoncxx::array::value>,
+        UniqueGenerator<bsoncxx::document::value>,
+        UniqueGenerator<bsoncxx::types::b_date>,
+        UniqueGenerator<double>,
+        UniqueGenerator<int64_t>,
+        UniqueGenerator<std::string>
+    >;
+
+    DistinctArrayGenerator(
+        const Node& node, GeneratorArgs generatorArgs, GeneratorTypes&& valueGen)
+        : _rng{generatorArgs.rng},
+          _node{node},
+          _generatorArgs{generatorArgs},
+          _valueGen{std::move(valueGen)},
+          _nTimesGen{intGenerator(extract(node, "number", "^DistinctArray"), generatorArgs)} {}
+
+    bsoncxx::array::value evaluate() override {
+        bsoncxx::builder::basic::array builder{};
+        const auto times = _nTimesGen->evaluate();
+        // Save some time if we don't want to have any items.
+        if (times < 1) {
+            return builder.extract();
+        }
+
+        // Use a visitor pattern to specialize the filtering.
+        std::visit([&] (auto&& generator) {
+            // Extract the first value so that we can use it to do type deduction for the set.
+            const auto&& firstValue = generator->evaluate();
+            // Choose initial bucket size of 2N to minimize hash collisions and resizes.
+            std::unordered_set distinctValues({firstValue}, 2 * times, setHasher, setKeyEq);
+            // Choose a buffer length of max(times, 64).
+            const auto bufferLength = std::max<int64_t>(64, times);
+            // Keep track of collision offsets for next N runs.
+            std::deque<int8_t> offsets(bufferLength);
+            ssize_t collisions = 0;
+
+            /**
+             * Sliding window via cyclic buffer.
```

Comment:
uhhh... helper function/class with tests maybe? This is pretty complex to inline for this purpose... Can we just say "failure to insert 100 times in a row is fatal?" That's probably more than good enough for our purposes

Reply:
I just worry that debugging this cyclic buffer may be harder than detecting that you have induced an infinite loop. The error message is also pretty... scary? None of those words are easy to parse/associate with "infinite loop"